### ISSUE TITLE
Loading dependencies is not thread safe

### ISSUE THREAD

**Issue by flash-gordon** [dry-rb/dry-system#149]
**Describe the bug**

TLDR; I wrote a spec (in my project) that spins two threads that load dependencies. Obviously, I don't load the whole container in tests and hence I get two threads loading the same components _concurrently_. This shouldn't be a problem but in fact, I get weird loading errors.

**To Reproduce**

This is something to work on. I use dry-system with dry-effects but I reckon it shouldn't matter. I'll try to come up with a simple repro (and a fix). I already have a fix so writing a failing spec is the next step.

**Expected behavior**

dry-system must be resilient to concurrent and parallel loading. It's fine to preload the whole app in production but the gem also designed with the idea "load what you need". Threads/fibers/whatnot shouldn't make a difference.

**Your environment**

- Affects my production application: **I WORKED AROUND**
- Ruby version: N/A
- OS: N/A

**Comment by timriley** [comment]
@flash-gordon Do you have more details about this? Or that reproduction, by any chance? As we get close to 1.0.0 this is obviously an important one to fix.

**Comment by flash-gordon** [comment]
Oops, I forgot about this. IIRC I added a mutex around loader in my code. Also, I failed to make a standalone repro for the issue. I'll check on this today or tomorrow.

**Comment by timriley** [comment]
@flash-gordon thank you!

**Comment by flash-gordon** [comment]
At the moment, I have nothing to prove there's a problem. It's not clear to me how to reproduce the issue. It's possible to write a complex test that starts up a dry-system app in a threaded environment. But it's hard to tell if it'll be failing in advance. I don't have time to write it right now, unfortunately. The fix I have though is quite trivial, if someone shows up with a repro it'll be easy to check if the fix works for their case.

**Comment by flash-gordon** [comment]
For the record, I commented out the fix, the app didn't start failing after that 🤷  I guess, the issue is more subtle

**Comment by timriley** [comment]
Thanks for looking into this again, @flash-gordon! I might close this issue for now, then, since I'm sure we'll make sure it's known if any of us encounter this in the future.
